lib/Dialect/RTL/Ops.cpp

```diff
@@ -170,18 +172,42 @@ OpFoldResult AndOp::fold(ArrayRef<Attribute> operands) {
       value.isNullValue())
     return inputs()[size - 1];
 
-  /// TODO: and(..., '1) -> and(...) -- identity
-  /// TODO: and(..., x, x) -> and(..., x) -- idempotent
-  /// TODO: and(..., c1, c2) -> and(..., c3) where c3 = c1 & c2 -- constant
-  /// folding
-  /// TODO: and(x, and(...)) -> and(x, ...) -- flatten
-  /// TODO: and(..., x, not(x)) -> and(..., 0) -- complement
-
   return {};
 }
 
+void AndOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
+                                        MLIRContext *context) {
+  struct Folder final : public OpRewritePattern<AndOp> {
+    using OpRewritePattern::OpRewritePattern;
+    LogicalResult matchAndRewrite(AndOp op,
+                                  PatternRewriter &rewriter) const override {
+      auto inputs = op.inputs();
+      auto size = op.inputs().size();
+      APInt value;
+
+      // and(..., '1) -> and(...) -- identity
+      if (matchPattern(inputs[size - 1], m_RConstant(value)) &&
+          value.isAllOnesValue()) {
```

Comment:
This should check that `size() > 1` to avoid making a zero operand `And`

Reply:
We get here after verifier, that will check `(size < 1)` and folder, that will fold `(size == 1)`
So we should not have `(size < 2)` cases at this point.
Do you want to put an assertion here?

Reply:
oh right, if folders are guaranteed to run first, then an assertion would be great.

Reply:
added `size > 1` assertion